File: 01-intro/03-basic-combinatorial-algorithms/01-permutations/linear-approach.ex
defmodule PermutationCalculator do
  @size 3

  def print_buffer(buffer) do
    buffer |> Map.values() |> Enum.map(fn x -> x + 1 end) |> IO.inspect()
  end

  def permutate(index, used, buffer) do
    if index >= @size do
      print_buffer(buffer)
    end

    0..(@size - 1)
    |> Enum.filter(fn subindex -> not Map.get(used, subindex) end)
    |> Enum.each(fn subindex ->
      permutate(index + 1, %{used | subindex => true}, %{buffer | index => subindex})
    end)
  end

  def permutate() do
    used = 0..(@size - 1) |> Enum.map(fn x -> {x, false} end) |> Map.new()
    buffer = 0..(@size - 1) |> Enum.map(fn x -> {x, nil} end) |> Map.new()
    permutate(0, used, buffer)
  end
end

PermutationCalculator.permutate()

File: 01-intro/03-basic-combinatorial-algorithms/01-permutations/linear-approach.php
<?php

class PermutationCalculator
{
    protected $size = 3;
    protected $used = [];
    protected $buffer = [];

    public function __construct(int $size = 3)
    {
        $this->size = $size;
        $this->reset();
    }

    public function reset()
    {
        for ($index = 0; $index < $this->size; $index++)
        {
            $this->used[$index] = false;
        }
    }

    public function print_buffer()
    {
        foreach ($this->buffer as $key => $value)
        {
            echo ($value + 1) . " ";
        }
        echo "\n";
    }

    public function permutate(int $index = 0)
    {
        if ($index >= $this->size) {
            $this->print_buffer();
            return;
        }

        for ($i = 0; $i < $this->size; $i++)
        {
            if (!$this->used[$i]) {
                $this->used[$i] = true;
                $this->buffer[$index] = $i;
                $this->permutate($index + 1);
                $this->used[$i] = false;
            }
        }
    }
}

echo (new PermutationCalculator)->permutate();